basis/basis.py

```diff
@@ -0,0 +1,172 @@
+import abc
+import numpy
+import math
+
+class BasisFunction(abc.ABC):
+    """Abstract class for basis functions.
+
+    Creates a framework for basis functions that are the building blocks
+    for the terms in the surrogate model. 
+    ``max_exactness`` represents the maximum level of exactness, or 
+    accuracy, of a surrogate function this class can provide enough 
+    information for. This class can provide enough information for 
+    constructing a surrogate function of any accuracy at or below 
+    max_exactness.
+    ``points`` property represents all the unique 1D points of the 
+    basis function associated with a Smolyak index described by the 
+    class IndexGrid. The Smolyak index of a given point is equal to 
+    its position in the list points.
+    ``levels`` property gives the Smolyak indexes for each
+    grid level.
+
+    Parameters
+    ----------
+    max_exactness : int
+        level of exactness class will describe
+
+    """
+
+    def __init__(self,max_exactness):
```

Comment:
After our meeting last week, I don't think the base class `BasisFunction` should have `max_exactness` as a constructor argument. Not all basis functions need to have this.